### PR TITLE
Add `platform` filtering option to `graph` command

### PR DESCRIPTION
**Pull request by mikchmie**
### Short description 📝

Since the "multiplatform support for external SPM dependencies" (#4570) has been added in 3.9.0 release, graphs produced by `tuist graph` command have become harder to read in projects with big dependency trees. This is because all libraries supporting multiple platforms are represented as multiple targets on the graph (one per platform).
There's even a warning emitted by the command, indicating that the graph is too big to be rendered in full quality:
```
dot: graph is too large for cairo-renderer bitmaps. Scaling by 0.775275 to fit
```

This PR aims to mitigate the issue by allowing to limit the amount of targets in the graph, by filtering output by specific platform.

Here's a rough example of a graph before and after filtering to a single platform:
#### Full graph
Full graph

#### Filtered graph
Filtered graph

### How to test the changes locally 🧐

1. Create a project with a multi-platform dependency (e.g. `Alamofire`)
2. `tuist graph --platform ios`

Expected: Graph contains only iOS targets (e.g. includes `Alamofire_ios`, but not `Alamofire_macos`)

### Checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase.
- [ ] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [ ] The PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, whenever it should be included in the “Added”, “Fixed” or “Changed” section of the CHANGELOG. Note: when included in the CHANGELOG, the title of the PR will be used as entry, please make sure it is clear and suitable.
- [ ] In case the PR introduces changes that affect users, the documentation has been updated.
- [ ] In case the PR introduces changes that affect how the cache artifact is generated starting from the `TuistGraph.Target`, the `Constants.cacheVersion` has been updated.

### CODE REVIEW

**Comment by mikchmie** [comment]
I'm open for any suggestions regarding CLI. The new syntax looks like this after my changes:
```
tuist graph [--skip-test-targets] [--skip-external-dependencies] [--platforms <platforms> ...] [--format <format>] [--no-open] [--algorithm <algorithm>] [<targets> ...] [--path <path>] [--output-path <output-path>]
```

I see one potential problem with my proposal, but I think it's negligible. It might be hard to specify `targets` arguments together with `platforms` option, as targets will be treated as input to `platforms` option.
```
tuist graph --platforms ios watchos MyApp MyWatchApp
                       |    ^      |       ^         |
                       | Platforms |    Targets      |
```
However I don't see any valid use case for filtering by targets and platforms at the same time.

**Review by danyf90 — CHANGES_REQUESTED** [review]
What about supporting a single platform as a filter? I don't expect filtering by more platform to be really needed, and it would simplify the API

**Comment by danyf90** [comment]
Also, remember to update the graph.md documentation file. I suggest you rebase on top of https://github.com/tuist/tuist/pull/4659 before doing it

**Comment by mikchmie** [comment]
> What about supporting a single platform as a filter? I don't expect filtering by more platform to be really needed, and it would simplify the API

@danyf90 I was thinking that e.g. for iOS + watchOS projects, it might be useful to filter by those 2 platforms (to not see macOS and tvOS dependencies).

**Comment by danyf90** [comment]
> > What about supporting a single platform as a filter? I don't expect filtering by more platform to be really needed, and it would simplify the API
> 
> 
> 
> @danyf90 I was thinking that e.g. for iOS + watchOS projects, it might be useful to filter by those 2 platforms (to not see macOS and tvOS dependencies).

If you have an iOS + watchOS project the macOS and tvOS are not generated

**Comment by mikchmie** [comment]
> If you have an iOS + watchOS project the macOS and tvOS are not generated

Could you elaborate on that? I've just added watchOS extension to my project (alongside an iOS app) and I see dependencies for all Apple platforms (including macOS and tvOS) on the graph.

**Comment by danyf90** [comment]
> > If you have an iOS + watchOS project the macOS and tvOS are not generated
> 
> Could you elaborate on that? I've just added watchOS extension to my project (alongside an iOS app) and I see dependencies for all Apple platforms (including macOS and tvOS) on the graph.

Have you specified the correct platforms in `Dependencies.swift`?

**Comment by mikchmie** [comment]
> Have you specified the correct platforms in `Dependencies.swift`?

Oh, I didn't know there's such an option 😬 Then I guess simply setting `iOS` there will solve my initial problem. 
And now I agree with you point of view - filtering by a single platform should be enough for someone looking to make the graph more readable in mulitplatform project.

File: Sources/TuistGenerator/Extensions/Graph+Extras.swift [hunk 1]
```diff
@@ -21,6 +22,10 @@ extension TuistGraph.Graph {
                 return false
             }
 
+            if let platformToFilter = platformToFilter, target.target.platform != platformToFilter {
```
Comment (danyf90) [inline Sources/TuistGenerator/Extensions/Graph+Extras.swift:25]:
```suggestion
            if target.target.platform != platformToFilter {
```

**Comment by mikchmie** [comment]
I'll update the documentation once #4659 is merged.

**Review by danyf90 — APPROVED** [review]

File: Sources/TuistGenerator/Extensions/Graph+Extras.swift (hunk 1 above)
Comment (mikchmie) [reply Sources/TuistGenerator/Extensions/Graph+Extras.swift:25]: When `platformToFilter` is `nil`(the default value), nothing goes through that filter and we want the exact opposite behavior (everything goes through).

Comment (danyf90) [reply Sources/TuistGenerator/Extensions/Graph+Extras.swift:25]: right, sorry :D
I would still use a guard as it's a precondition for going on:
```
guard platformToFilter == nil || target.target.platform == platformToFilter else {
```

**Pull request merged**